File: leetcode/List/25.ReverseNodesInk-Group/25ReverseNodesInk-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

###############################
############solution1##########

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        stack = []
        dummy = ListNode(0)
        dummy.next = head
        current = dummy
        curr_next = current.next
        while curr_next:
            for i in range(k):
                if curr_next:
                    stack.append(curr_next)
                    curr_next = curr_next.next
            #end_for
            if len(stack) != k:
                return dummy.next
            while len(stack):
                current.next = stack.pop()
                current = current.next
                print(current.val)
            #end
            current.next = curr_next
        #end_while
        
        return dummy.next


####################################
################solution2###########

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return None
        dummy = ListNode(0)
        dummy.next = head
        pre = dummy
        
        def reverse(pre: ListNode, k: int) -> ListNode:
            last = pre
            for i in range(k+1):
                last = last.next
                if i != k and last is None:
                    return None
            # 对pre  和  last 中间的节点做reverse
            tail = pre.next
            curr = pre.next.next
            while curr != last:
                curr_next = curr.next
                curr.next = pre.next
                pre.next = curr
                tail.next = curr_next
                curr = curr_next
            #end_while
            return tail
        ###
        while pre:
            pre = reverse(pre, k)
        return dummy.next

File: leetcode/List/25.ReverseNodesInk-Group/25ReverseNodesInk-Group.js
/*
 * @lc app=leetcode.cn id=25 lang=javascript
 *
 * [25] K 个一组翻转链表
 */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    var temp = head;
    for (var i = 1; i < k && temp !== null; i++) {
        temp = temp.next;
    }
    if (temp == null) return head;
    var t2 = temp.next;
    temp.next = null;
    var newHead = reverseList(head);
    var newTemp = reverseKGroup(t2, k);

    head.next = newTemp;
    return newHead;
};

function reverseList(head) {
    if (head == null || head.next == null) return head;
    var res = reverseList(head.next);
    head.next.next = head;
    head.next = null;
    return res;
}
